### ISSUE TITLE
Horizontal spacing commands: \>, \hfill

### ISSUE THREAD

**Issue by xymostech** [Khan/KaTeX#164]
We have some spacing commands, but it appears we missed one. There might be more that we missed.

**Comment by kevinb7** [comment]
We currently support: ```\!```, ```\ ```, ```\,```, ```\;```, ```\:```, ```~```, ```\space```, ```\enspace```, ```\quad```, ```\qquad```, and ```\phantom```.

We're missing: ```\thinspace```, ```\hspace```, ```\hfill```, ```\kern```, and ```\hphantom```.

Notes:
- ```\thinspace``` is an alias for ```\,```
- ```\hphantom``` is not an alias for ```\phantom```, according to TeXBook pg 178:

> There’s also \hphantom, which has the width of a \phantom, but its
height and depth are zero.

Reference: http://tex.stackexchange.com/questions/74353/what-commands-are-there-for-horizontal-spacing

**Comment by gagern** [comment]
The way I see it, `\kern` is often followed by a dimension without braces around it. Would we want to support that as well? If so, we would have to adjust the [parsing of sizes in my macros branch](https://github.com/gagern/KaTeX/blob/b81fcb8765bbdb77c0514c566d651a8726dde15f/src/Parser.js#L609-L617), at least if we actually follow some variant of that branch to implement macros.

**Comment by kevinb7** [comment]
I guess we'd have to.  Should probably have a ```lexSize``` function.

**Comment by gagern** [comment]
Nah, we shouldn't do this in the lexer, but on the token stream. That's what TeX does, as far as I understand things.

**Comment by kevinb7** [comment]
That could work to.  It seems like using a regex directly on the input stream would be easier as opposed to having to concatenate tokens would be trickier.  I know there's code there that does this already, but it relies on a hard stop which we wouldn't have in this situation.

**Comment by gagern** [comment]
I think a length would be finished after its two letter unit has been parsed. But I'd have to read The TeX Book to verify this first.

**Comment by kevinb7** [comment]
> I think a length would be finished after its two letter unit has been parsed.

For sure, it's a little trickier than checking for a ```"]"```.

**Comment by gagern** [comment]
True, but it has the benefit of working even if there is some macro expansion involved in the length notation.

**Comment by michaelcadilhac** [comment]
Any update on that?  It seems that `hspace` is a reasonable thing to have indeed.  Cheers.

**Comment by kevinbarabash** [comment]
@michaelcadilhac we do have macros now, and we do support `\kern` now

**Comment by michaelcadilhac** [comment]
@kevinbarabash certainly, hence why not hspace?

**Comment by kevinbarabash** [comment]
@michaelcadilhac just haven't gotten it, mainly busy with other stuff and there's a pile of almost-there PRs that I should finish first.  PRs are welcome though.

**Comment by edemaine** [comment]
Is there a difference between `\hspace` and `\kern`, in either math or text mode?  If not, I can throw together a macro-based PR.  [Here's a list of various horizontal spacing commands](https://tex.stackexchange.com/questions/74353/what-commands-are-there-for-horizontal-spacing), which we should review for support.

**Comment by edemaine** [comment]
OK, so here's the difference: `\kern` supports units `mu` and `ex` only, and no `plus` or `minus` arguments, whereas `\hskip`/`\hspace` supports lots of units and `plus`/`minus` arguments.  I don't see `plus`/`minus` as important in KaTeX (it's not like we have boxes whose width we can adjust), but supporting other units would be important.  

Incidentally, most other commands listed above work fine in KaTeX, including the very useful `\quad`, `\qquad`, and `~`.

**Comment by gagern** [comment]
`plus`, `minus` and stretchable glue *might* become relevant at some point if we manage to use flex boxes to model these. Should be possible, at least in theory and for some situations. Not sure how urgent this is, so I'd say we shouldn't block on support for it.

**Comment by Casyfill** [comment]
oh, waiting is hard... :-)

**Comment by edemaine** [comment]
@Casyfill If you want `\hspace` to work now, try defining a macro mapping it to `\kern`.  `\kern` is essentially behaving like `\hspace` now, and will work with units of `em` and `ex`.  `\kern` *should* behave differently, and I'll work on a PR to that effect, but this is a good stopgap for now.

We're also working on unit support on #732, which is still a work in progress, and has some discussion about best way to do it -- feel free to join in.

**Comment by kevinbarabash** [comment]
The two remaining commands are `\>` and `\hfill`.

**Comment by edemaine** [comment]
To clarify, #833 added basic support for `\hspace` (everything except `plus` and `minus`). I'm also looking at supporting `\hspace*` (as an alias to `\hspace`).

**Comment by kevinbarabash** [comment]
Closing this in favour of #990 since all horizontal spacing commands other than `\hfil` et al have been implemented.
